pkg/gateway/server/llmproxy.go

```diff
@@ -241,3 +310,76 @@ func (r *responseModifier) Close() error {
 	}
 	return r.c.Close()
 }
+
+func mustParseURL(s string) *url.URL {
+	u, err := url.Parse(s)
+	if err != nil {
+		panic(err)
+	}
+	return u
+}
+
+type llmProviderProxy struct {
+	dailyUserTokenPromptTokenLimit     int
+	dailyUserTokenCompletionTokenLimit int
+	u                                  url.URL
+	modelProviderName                  string
+	modelProvider                      *v1.ToolReference
+	lock                               sync.RWMutex
+}
+
+func (s *Server) newLLMProviderProxy(u *url.URL, modelProviderName string) *llmProviderProxy {
+	return &llmProviderProxy{
+		dailyUserTokenPromptTokenLimit:     s.dailyUserTokenPromptTokenLimit,
+		dailyUserTokenCompletionTokenLimit: s.dailyUserTokenCompletionTokenLimit,
+		u:                                  *u,
+		modelProviderName:                  modelProviderName,
+	}
+}
+
+func (l *llmProviderProxy) proxy(req api.Context) error {
+	l.lock.RLock()
+	modelProvider := l.modelProvider
+	l.lock.RUnlock()
+
+	if modelProvider == nil {
+		modelProvider = new(v1.ToolReference)
+		if err := req.Get(modelProvider, l.modelProviderName); err != nil {
+			return fmt.Errorf("model provider %s not found: %w", l.modelProviderName, err)
+		}
+
+		l.lock.Lock()
+		l.modelProvider = modelProvider
+		l.lock.Unlock()
+	}
+
+	remainingUsage, err := req.GatewayClient.RemainingTokenUsageForUser(req.Context(), req.User.GetUID(), tokenUsageTimePeriod, l.dailyUserTokenPromptTokenLimit, l.dailyUserTokenCompletionTokenLimit)
+	if err != nil {
+		return err
+	} else if !remainingUsage.UnlimitedPromptTokens && remainingUsage.PromptTokens <= 0 || !remainingUsage.UnlimitedCompletionTokens && remainingUsage.CompletionTokens <= 0 {
+		return types2.NewErrHTTP(http.StatusTooManyRequests, fmt.Sprintf("no tokens remaining (prompt tokens remaining: %d, completion tokens remaining: %d)", remainingUsage.PromptTokens, remainingUsage.CompletionTokens))
```

Comment:
Why do we only error when they are out of both prompt and completion tokens? Shouldn't we error if outOfPrompt || outOfCompletion?

Reply:
Based on the logic here, I believe that is what we are doing: (if the user doesn't have unlimited prompt tokens AND they don't have any prompt tokens left) OR (the user doesn't have unlimited completion tokens AND they don't have any completion tokens left)

Reply:
Ah sorry, I misread the code